File: solutions/230-kth-smallest-element-in-a-bst/solution.go
package main

import (
	"fmt"
	"strings"
	"strconv"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTreeFromLevel(values []interface{}) *TreeNode {
	if len(values) == 0 || values[0] == nil {
		return nil
	}
	root := &TreeNode{Val: values[0].(int)}
	queue := []*TreeNode{root}
	i := 1
	for len(queue) > 0 && i < len(values) {
		node := queue[0]
		queue = queue[1:]
		if i < len(values) && values[i] != nil {
			node.Left = &TreeNode{Val: values[i].(int)}
			queue = append(queue, node.Left)
		}
		i++
		if i < len(values) && values[i] != nil {
			node.Right = &TreeNode{Val: values[i].(int)}
			queue = append(queue, node.Right)
		}
		i++
	}
	return root
}

func parseTreeLevel(s string) []interface{} {
	s = strings.Trim(s, "[]")
	if s == "" {
		return nil
	}
	parts := strings.Split(s, ",")
	out := make([]interface{}, 0, len(parts))
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p == "null" || p == "" {
			out = append(out, nil)
		} else {
			v, _ := strconv.Atoi(p)
			out = append(out, v)
		}
	}
	return out
}

// 思路：中序遍历，记录遍历的节点数量，当遍历到第k个节点时，返回该节点的值
func kthSmallest(root *TreeNode, k int) int {
	count := 0
	var result int
	var dfs func(node *TreeNode)
	dfs = func(node *TreeNode) {
		if node == nil {
			return
		}
		dfs(node.Left)
		count++
		if count == k {
			result = node.Val
			return
		}
		dfs(node.Right)
	}
	dfs(root)
	return result
}

func main() {
	root := buildTreeFromLevel(parseTreeLevel("3,1,null,2,4"))
	fmt.Printf("PASS | %d\n", kthSmallest(root, 1))
}

File: solutions/230-kth-smallest-element-in-a-bst/solution.py
# 思路：中序遍历，记录遍历的节点数量，当遍历到第k个节点时，返回该节点的值
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def kth_smallest(self, root: Optional[TreeNode], k: int) -> int:
        count = 0
        result = 0

        def dfs(node: Optional[TreeNode]) -> None:
            nonlocal count, result
            if node is None:
                return
            dfs(node.left)
            count += 1
            if count == k:
                result = node.val
                return
            dfs(node.right)

        dfs(root)
        return result


if __name__ == "__main__":
    import sys
    from pathlib import Path
    sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "_helpers"))
    from testutil import build_tree

    sol = Solution()
    root = build_tree([3, 1, 4, None, 2])
    got = sol.kth_smallest(root, 1)
    status = "PASS" if got == 1 else "FAIL"
    print(f"{status} | kth={got}")
